libraries/lts/include/mcrl2/lts/detail/liblts_coupledsim.h

```diff
@@ -0,0 +1,592 @@
+      {
+        // pop and keep just start and extension.
+        // finish if next is second not tau.
+        transition weak = todo_weak.top();
+        todo_weak.pop();
+        size_t f = weak.from();
+        size_t l = weak.label();
+        size_t t = weak.to();
+        bool already_good = !l2.is_tau(l);  // path already has a good action
+
+        std::map<transition,bool> next = l2_tran_from_node[t];
+        size_t len = next.size();
+
+        if (true)  // all, also just tau chains may be requeseted later
+          // if (already_good)  // (actually already) done
+        {
+          /* The current todo weak transition is already valid.*/
+          // if it was strong before, it stays strong, else added as weak.
+          l2_weak_transitions.insert(weak);
+          l2_tran_into_node[t][weak] |= false;
+          l2_tran_from_node[f][weak] |= false;
+        }
+
+        if (len < 1)  // no further steps.
+        {
+          continue;
+        }
+        else  // just extend simply.
+        {
+          for (const auto &ntrans : next)
+          {
+            size_t next_label = ntrans.first.label();
+            bool next_tau = l2.is_tau(next_label);
+
+            /* If tau: extend new todo with extension.
+             * If all before only tau: extend new todo with extension.
+             */
+            if (next_tau || !already_good)
+            {
+              /* Maybe use new label: If now good.*/
+              transition new_extended_weak
+                = transition(f, !already_good ? next_label : l, ntrans.first.to());
+
+              // re-add new branches.
+              todo_weak.push(new_extended_weak);
+            }
+          }
+        }
+        // cuurent weak transition is done now.
+      }  // done l2 tau forest (all tau pathes).
+    }
+
+    mCRL2log(log::verbose)
+      << "Creating now the cs-game arena."
+      << std::endl;
+
+    for (size_t p0 = 0; p0 < l1.num_states(); p0++)
+    {
+      for (size_t q0 = 0; q0 < l2.num_states(); q0++)
+      {
+        cs_game_node p0q0 = {NODE_ATK, 0, p0, q0, false};  // atk (p0,q0)
+        cs_game_node cplp0q0 = {NODE_CPL, 0, p0, q0, false}; // (cpl,p0,q0)
+
+        /* swapped. */
+        cs_game_node q0p0 = {NODE_ATK, 0, q0, p0, true};  // swapped (q0,p0)
+        cs_game_node cplq0p0 = {NODE_CPL, 0, q0, p0, true};  // swapped (cpl,q0,p0)
+
+        attacker_nodes.insert(p0q0);
+        attacker_nodes.insert(q0p0);
+        defender_nodes.insert(cplp0q0);
+        defender_nodes.insert(cplq0p0);
+
+        moves.insert({p0q0, cplp0q0, 0, "cpl"});  // (p0,q0) -> (Cpl,p0,q0)
+        moves.insert({q0p0, cplq0p0, 0, "cpl"});  // (q0,p0) -> (Cpl,q0,p0)
+
+        /* bisim: coupling answer q'=q, p'=p*/
+        moves.insert({cplp0q0, q0p0, 0, "bisim"});
+        moves.insert({cplq0p0, p0q0, 0, "bisim"});
+
+        std::map<cs_game_move, bool> todo_if;
+
+        // TODO this includes also weak, as challenge giver invalid, solve!
```

Comment:
Does this TODO item mean that the implemented algorithm is not yet fully correct? If so, what steps are necessary in order to finalise the algorithm. Same remark on line 419.